File: scripts/capture-live.mjs
import { mkdir, writeFile } from "node:fs/promises";
import { dirname, resolve } from "node:path";

const currentUrl = process.env.CURRENT_RPC_URL;
const oldUrl = process.env.OLD_RPC_URL;
if (!currentUrl || !oldUrl) throw new Error('Set CURRENT_RPC_URL and OLD_RPC_URL before a live capture');
const output = resolve(process.env.OUTPUT ?? "evidence/live");

async function rpc(url, method, params) {
  const response = await fetch(url, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ jsonrpc: "2.0", id: 1, method, params }),
  });
  if (!response.ok) throw new Error(`${url} returned HTTP ${response.status}`);
  return response.json();
}

function channelKeys(payload) {
  return [...new Set((payload?.result?.channels ?? []).flatMap((channel) => Object.keys(channel)))].sort();
}

const requests = {
  node_info: [],
  list_channels_pending: [{ pubkey: null, include_closed: false, only_pending: true }],
};

const evidence = { captured_at: new Date().toISOString(), endpoints: {} };
for (const [label, url] of [["current", currentUrl], ["old", oldUrl]]) {
  evidence.endpoints[label] = { url, responses: {} };
  for (const [name, params] of Object.entries(requests)) {
    evidence.endpoints[label].responses[name] = await rpc(url, name === "list_channels_pending" ? "list_channels" : name, params);
  }
}

const currentChannels = evidence.endpoints.current.responses.list_channels_pending;
const oldChannels = evidence.endpoints.old.responses.list_channels_pending;
const currentKeys = channelKeys(currentChannels);
const oldKeys = channelKeys(oldChannels);
const sdkChannelKeys = [
  "channel_id", "is_public", "channel_outpoint", "pubkey", "funding_udt_type_script",
  "state", "local_balance", "offered_tlc_balance", "remote_balance", "received_tlc_balance",
  "latest_commitment_transaction_hash", "created_at", "enabled", "tlc_expiry_delta",
  "tlc_fee_proportional_millionths", "shutdown_transaction_hash",
];
const observedKeys = [...new Set([...currentKeys, ...oldKeys])].sort();

evidence.analysis = {
  current_channel_keys: currentKeys,
  old_channel_keys: oldKeys,
  sdk_declared_channel_keys: sdkChannelKeys.sort(),
  live_keys_missing_from_fiber_js: observedKeys.filter((key) => !sdkChannelKeys.includes(key)),
  version_difference_detected: evidence.endpoints.current.responses.node_info.result?.version !== evidence.endpoints.old.responses.node_info.result?.version,
  evidence_status: "live_rpc_capture",
};

await mkdir(dirname(output), { recursive: true });
await writeFile(`${output}/responses.json`, `${JSON.stringify(evidence, null, 2)}\n`);
await writeFile(`${output}/report.json`, `${JSON.stringify(evidence.analysis, null, 2)}\n`);
console.log(JSON.stringify(evidence.analysis, null, 2));

File: scripts/capture-live.py
#!/usr/bin/env python3
import json
import os
from datetime import datetime, timezone
from pathlib import Path
from urllib.request import Request, urlopen

CURRENT = os.environ.get("CURRENT_RPC_URL")
OLD = os.environ.get("OLD_RPC_URL")
if not CURRENT or not OLD:
    raise SystemExit("CURRENT_RPC_URL and OLD_RPC_URL must be set for a live capture")
OUTPUT = Path(os.environ.get("OUTPUT", "evidence/live"))
SDK_CHANNEL_TS = Path(os.environ.get("FIBER_JS_CHANNEL_TS", ""))

def rpc(url, method, params):
    body = json.dumps({"jsonrpc": "2.0", "id": 1, "method": method, "params": params}).encode()
    request = Request(url, data=body, headers={"content-type": "application/json"})
    with urlopen(request, timeout=10) as response:
        return json.load(response)

responses = {}
for label, url in (("current", CURRENT), ("old", OLD)):
    responses[label] = {
        "url": url,
        "node_info": rpc(url, "node_info", []),
        "list_channels_pending": rpc(
            url, "list_channels", [{"pubkey": None, "include_closed": False, "only_pending": True}]
        ),
    }

def channel_keys(payload):
    return sorted({key for channel in payload.get("result", {}).get("channels", []) for key in channel})

current_keys = channel_keys(responses["current"]["list_channels_pending"])
old_keys = channel_keys(responses["old"]["list_channels_pending"])
if not SDK_CHANNEL_TS:
    raise SystemExit("FIBER_JS_CHANNEL_TS must point to the pinned fiber-js Channel type used for this capture")
sdk_source = SDK_CHANNEL_TS.read_text()
channel_interface = sdk_source.split("interface Channel {", 1)[1].split("}\n", 1)[0]
sdk_keys = sorted({line.strip().split("?", 1)[0].split(":", 1)[0] for line in channel_interface.splitlines() if ":" in line})
observed_keys = sorted(set(current_keys + old_keys))
analysis = {
    "captured_at": datetime.now(timezone.utc).isoformat(),
    "current_channel_keys": current_keys,
    "old_channel_keys": old_keys,
    "sdk_declared_channel_keys": sdk_keys,
    "live_keys_missing_from_fiber_js": [key for key in observed_keys if key not in sdk_keys],
    "runtime_versions_differ": (
        responses["current"]["node_info"].get("result", {}).get("version")
        != responses["old"]["node_info"].get("result", {}).get("version")
    ),
    "observable_channel_shape_difference": current_keys != old_keys,
    "sdk_contract_gap_detected": bool([key for key in observed_keys if key not in sdk_keys]),
    "version_regression_claimed": False,
    "evidence_status": "live_rpc_capture",
}
OUTPUT.mkdir(parents=True, exist_ok=True)
(OUTPUT / "responses.json").write_text(json.dumps({"endpoints": responses, "analysis": analysis}, indent=2) + "\n")
(OUTPUT / "report.json").write_text(json.dumps(analysis, indent=2) + "\n")
print(json.dumps(analysis, indent=2))
